plugins/notifications-backend/src/service/NotificationCleaner.ts

```
@@ -0,0 +1,80 @@
+import { Config, readDurationFromConfig } from '@backstage/config';
+import { NotificationsStore } from '../database';
+import { durationToMilliseconds, HumanDuration } from '@backstage/types';
+
+export class NotificationCleaner {
+  static readonly initNotificationCleaner = async (
+    config: Config,
+    scheduler: SchedulerService,
+    logger: LoggerService,
+    database: NotificationsStore,
+  ): Promise<void> => {
+    const schedule: SchedulerServiceTaskScheduleDefinition = config.has(
+      'notifications.expiration.schedule',
+    )
+      ? readSchedulerServiceTaskScheduleDefinitionFromConfig(
+          config.getConfig('notifications.expiration.schedule'),
+        )
+      : {
+          frequency: { cron: '0 0 * * *' },
+          timeout: { hours: 1 },
+          initialDelay: { hours: 1 },
+          scope: 'global',
+        };
+
+    const expiration: HumanDuration = config.has(
+      'notifications.expiration.duration',
+    )
+      ? readDurationFromConfig(config, {
+          key: 'notifications.expiration.duration',
+        })
+      : { years: 1 };
+    const expirationMs = durationToMilliseconds(expiration);
+    if (expirationMs > 0) {
+      const taskRunner = scheduler.createScheduledTaskRunner(schedule);
+      await taskRunner.run({
+        id: 'notification-cleaner',
+        fn: async () => {
+          await this.clearNotifications(logger, database, expirationMs);
+        },
+      });
+    }
+  };
+
+  static readonly clearNotifications = async (
+    logger: LoggerService,
+    database: NotificationsStore,
+    expirationMs: number,
+  ) => {
+    logger.info('Starting notification cleaner task');
+    try {
+      const result = await database.clearNotifications(expirationMs);
+      logger.info(
+        `Notification cleaner task completed successfully, removed ${result} notifications`,
+      );
+    } catch (error) {
+      logger.error(`Notification cleaner task failed: ${error}`);
```

@freben would you suggest to re-throw `ForwardedError` with context instead of logging here?

---

I think the task logger will handle it then, yes